### ISSUE TITLE
[icetray] allow I3Module to take a parameter value that means use the default (Trac #2380)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix4#2372]
There has been a proliferation of tray segments that include default values in their parameters, these default values are often different than the default in the I3Module that they call. This is bad because often it can be difficult to see where a value comes from. 

The best solution would be a true optional, but python doesn't really support that. Some segments set their arguments to None and then if they are not None add them to the I3Module using the keyword argument trick. This is good but it would be nice if I3Module supported a way to do this natively. 

This can be implemented easily by having I3Modules check every incoming parameter against the python value of None and just set it to the default if that happens. I think that python's None is correct but some sort of global singleton called `I3Default` might be a better idea. 

<details>
 <summary><em>Migrated from <a href="https://code.icecube.wisc.edu/projects/icecube/ticket/2380">https://code.icecube.wisc.edu/projects/icecube/ticket/2380</a>, reported by kjmeagherand owned by olivas</em></summary>
<p>

```json
{
    "status": "closed",
    "changetime": "2020-08-05T15:44:12",
    "_ts": "1596642252249407",
    "description": "There has been a proliferation of tray segments that include default values in their parameters, these default values are often different than the default in the I3Module that they call. This is bad because often it can be difficult to see where a value comes from. \n\nThe best solution would be a true optional, but python doesn't really support that. Some segments set their arguments to None and then if they are not None add them to the I3Module using the keyword argument trick. This is good but it would be nice if I3Module supported a way to do this natively. \n\nThis can be implemented easily by having I3Modules check every incoming parameter against the python value of None and just set it to the default if that happens. I think that python's None is correct but some sort of global singleton called `I3Default` might be a better idea. \n",
    "reporter": "kjmeagher",
    "cc": "",
    "resolution": "fixed",
    "time": "2019-12-10T19:37:21",
    "component": "combo core",
    "summary": "[icetray] allow I3Module to take a parameter value that means use the default",
    "priority": "normal",
    "keywords": "",
    "milestone": "Autumnal Equinox 2020",
    "owner": "olivas",
    "type": "defect"
}
```

</p>
</details>

**Comment by icecube-trac** [comment]
Trac update at 20191210T19:44:54: jvansanten commented:

> This is not a bad idea, and could be implemented by adding a custom sentinel (e.g. `NOTHING = object()`), and special-casing `SetParameter()` to be a no-op on when given `NOTHING`.

**Comment by icecube-trac** [comment]
Trac update at 20191218T20:22:48: olivas changed milestone from "Winter Solstice 2019" to "Vernal Equinox 2020"

**Comment by icecube-trac** [comment]
Trac update at 20191218T20:22:54: 
* olivas changed status from "new" to "accepted"
* olivas changed owner from "" to "olivas"

**Comment by icecube-trac** [comment]
Trac update at 20200304T17:47:41: 
* olivas commented:

> Nice feature, but it's going to have to wait until this Summer at the earliest...pushing to the Fall just to be safe.


* olivas changed milestone from "Vernal Equinox 2020" to "Autumnal Equinox 2020"

**Comment by icecube-trac** [comment]
Trac update at 20200805T15:44:12: 
* kjmeagher changed resolution from "" to "fixed"
* kjmeagher commented:

> In changeset 181466:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="181466"
> Allow icetray modules to take a parameter called I3Default which will act as if no value was passed to this parameter. This is useful so that tray segments can have default values. 
> 
> Fixes #2372
> ```


* kjmeagher changed status from "accepted" to "closed"
